src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs

```diff
@@ -154,112 +154,45 @@ public Parser(KnownTypeSymbols knownSymbols)
-                        int tokenCount = tokenList.Count;
-
-                        bool isPartial = false;
-
-                        string[] declarationElements = new string[tokenCount + 2];
+                    bool isPartial = false;
 
-                        for (int i = 0; i < tokenCount; i++)
-                        {
-                            SyntaxToken token = tokenList[i];
-                            declarationElements[i] = token.Text;
+                    string[] declarationElements = new string[tokenCount + 2];
 
-                            if (token.IsKind(SyntaxKind.PartialKeyword))
-                            {
-                                isPartial = true;
-                            }
-                        }
+                    for (int i = 0; i < tokenCount; i++)
+                    {
+                        SyntaxToken token = tokenList[i];
+                        declarationElements[i] = token.Text;
 
-                        if (!isPartial)
+                        if (token.IsKind(SyntaxKind.PartialKeyword))
                         {
-                            classDeclarationList = null;
-                            return false;
+                            isPartial = true;
                         }
-
-                        declarationElements[tokenCount] = "class";
-                        declarationElements[tokenCount + 1] = GetClassDeclarationName(currentSymbol);
-
-                        (classDeclarationList ??= new List<string>()).Add(string.Join(" ", declarationElements));
                     }
 
-                    currentSymbol = currentSymbol.ContainingType;
-                }
-
-                Debug.Assert(classDeclarationList?.Count > 0);
-                return true;
-            }
-
-            private static string GetClassDeclarationName(INamedTypeSymbol typeSymbol)
-            {
-                if (typeSymbol.TypeArguments.Length == 0)
-                {
-                    return typeSymbol.Name;
-                }
-
-                StringBuilder sb = new StringBuilder();
-
-                sb.Append(typeSymbol.Name);
-                sb.Append('<');
-
-                bool first = true;
-                foreach (ITypeSymbol typeArg in typeSymbol.TypeArguments)
-                {
-                    if (!first)
+                    if (!isPartial)
                     {
-                        sb.Append(", ");
-                    }
-                    else
-                    {
-                        first = false;
+                        classDeclarationList = null;
+                        return false;
                     }
 
-                    sb.Append(typeArg.Name);
-                }
+                    INamedTypeSymbol typeSymbol = semanticModel.GetDeclaredSymbol(current, cancellationToken)!;
+                    declarationElements[tokenCount] = typeSymbol.IsValueType ? "struct" : "class";
```

Comment:
Do we already correctly handle other more specific kinds of types, like ref struct, record class, etc.?

Reply:
We don't, I opportunistically addressed the low-hanging fruit but it doesn't support arbitrary kinds.